src/lib/ext-activitylog.js

```diff
@@ -24,6 +25,36 @@ class Model {
       this.filter[logKey].splice(logKeyIndex, 1);
     }
   }
+
+  setFilterKeyword(keyword) {
+    this.filter.keyword = keyword;
+  }
+
+  matchLogWithFilterObj(log) {
+    return (
+      this.matchFilterId(log.id) ||
+      this.matchFilterViewType(log.viewType) ||
+      this.matchFilterType(log.type) ||
+      this.matchFilterKeyword(log.data)
+    );
+  }
+
+  matchFilterId(logId) {
+    return this.filter.id.includes(logId);
+  }
+
+  matchFilterViewType(logViewType) {
+    return this.filter.viewType.includes(logViewType);
+  }
+
+  matchFilterType(logType) {
+    return this.filter.type.includes(logType);
+  }
+
+  matchFilterKeyword(logData) {
+    const logDataStr = JSON.stringify(logData);
+    return !logDataStr.includes(this.filter.keyword);
```

Comment:
This logic looks inverted/incorrect.

Now, with all filtering logic moved to the model, it should be easier to write unit tests to verify that the filtering logic works as expected. Can you add such unit tests?

Reply:
This is inverted since I have to check from the opposite direction compared to other checkings above. 
Yeah, I can add test for filtering logic.

Reply:
Why do you want to check the opposite direction?

The log should be shown if the filter matched.

A potential issue of your current logic is that you're using `||`, i.e. showing logs if any of the filters matched. I believe that you should change it to `&&`, show if all filters match.

Reply:
It's actually hiding logs when any of the filters match (since I am returning the boolean for setting [`row.hidden`](https://github.com/mozilla/extension-activity-monitor/blob/master/src/lib/web-component/log-view/log-view-element.js#L30) property. Here, I am using the filter object to collect the items which I want to get rid off. 

For example, if the `viewType` key in filter object has value `popup` in it, that means, I am getting rid of the logs whose review type is `popup`.

If I use `&&`, I will also have to invert the check [here](https://github.com/mozilla/extension-activity-monitor/blob/master/src/lib/web-component/log-view/log-view-element.js#L30). Also, currently I am not writing into filter object initially, but in this alternative approach, I will have to store all the filter labels into the filter object and remove them when user unchecks the checkboxes. Should I do that?

Reply:
Logis should be `.hidden = !filter(log)`

`filter`'s contract is usually "true if the element should be shown", for example with Array's filter: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

Reply:
done :-)